uxarray/helpers.py

```diff
@@ -39,6 +51,372 @@ def parse_grid_type(filepath, **kw):
+        else:
+            msg = str(e) + ': {}'.format(filepath)
+    finally:
+        if msg != "":
+            msg = "Unable to determine file type, mesh file not supported" + ': {}'.format(
+                filepath)
+            raise ValueError(msg)
+
+    return mesh_filetype
+
+
+@njit
+def _spherical_to_cartesian_unit_(node, r=6371):
+    """Converts spherical (lat/lon) coordinates to cartesian (x,y,z).
+
+    Final output is cartesian coordinates on a sphere of unit radius
+
+    Parameters:
+    -----------
+    node: a list consisting of lat and lon
+
+    Returns: numpy array
+        Cartesian coordinates of length 3
+    """
+    lon = node[0]
+    lat = node[1]
+    lat, lon = np.deg2rad(lat), np.deg2rad(lon)
+    x = r * np.cos(lat) * np.cos(lon)  # x coordinate
+    y = r * np.cos(lat) * np.sin(lon)  # y coordinate
+    z = r * np.sin(lat)  # z coordinate
+
+    coord = np.array([x, y, z])
+    # make it coord on a sphere with unit radius
+    unit_coord = coord / np.linalg.norm(coord)
+
+    return unit_coord
+
+
+# Calculate the area of all faces.
+@njit
+def calculate_face_area(x,
+                        y,
+                        z,
+                        quadrature_rule="gaussian",
+                        order=4,
+                        coords_type="spherical"):
+    """Calculate area of a face on sphere.
+
+    Parameters
+    ----------
+    x : list, required
+        x-coordinate of all the nodes forming the face
+
+    y : list, required
+        y-coordinate of all the nodes forming the face
+
+    z : list, required
+        z-coordinate of all the nodes forming the face
+
+    quadrature_rule : str, optional
+        triangular and Gaussian quadrature supported, expected values: "triangular" or "gaussian"
+
+    order: int, optional
+        Order of the quadrature rule. Default is 4.
+
+        Supported values:
+            - Gaussian Quadrature: 1 to 10
+            - Triangular: 1, 4, 8, 10 and 12
+
+    coords_type : str, optional
+        coordinate type, default is spherical, can be cartesian also.
+    """
+    area = 0.0  # set area to 0
+    order = order
+
+    if quadrature_rule == "gaussian":
+        dG, dW = get_gauss_quadratureDG(order)
+    elif quadrature_rule == "triangular":
+        dG, dW = get_tri_quadratureDG(order)
+    else:
+        raise ValueError(
+            "Invalid quadrature rule, specify gaussian or triangular")
+
+    num_nodes = len(x)
+
+    # num triangles is two less than the total number of nodes
+    num_triangles = num_nodes - 2
+
+    # Using tempestremap GridElements: https://github.com/ClimateGlobalChange/tempestremap/blob/master/src/GridElements.cpp
+    # loop through all sub-triangles of face
+    for j in range(0, num_triangles):
+        node1 = np.array([x[0], y[0], z[0]], dtype=np.float64)
+        node2 = np.array([x[j + 1], y[j + 1], z[j + 1]], dtype=np.float64)
+        node3 = np.array([x[j + 2], y[j + 2], z[j + 2]], dtype=np.float64)
```

Comment:
This node selection may not always give us the correct triangles, so, is there an assumption about the order of the nodes in the original tempestremap code about these? For example, the below scenario won't give us the correct area:

<img width="187" alt="image" src="https://user-images.githubusercontent.com/32553057/179337755-d1b673d2-195d-4c8d-9090-6af385a4b535.png">

May there be something getting lost from tempestremap to here?

Reply:
> This node selection scheme may not always give us the correct triangles, so, is there an assumption about the order of the nodes in the original tempestremap code about these? For example, the below scenario won't give us the correct area:
> 
> <img alt="image" width="187" src="https://user-images.githubusercontent.com/32553057/179337755-d1b673d2-195d-4c8d-9090-6af385a4b535.png">
> 
> Am I missing anything here or may there be something getting lost from tempestremap to here?

This is not a valid mesh, there is a function called validate in our grid class. Once that is completed this mesh would be rejected. The assumption is: nodes are in counterclockwise order, and we look through the connectivity of the face breaking it into triangles. Incorrect order of nodes in connectivity will not produce the correct result

Reply:
Thanks for the explanation! This reminds me we may also consider putting a CCW ordering utility function into our package in the future?